File: regions/system/library/typescript/building-blocks/src/inmemory_pubsub.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { InMemoryPubSub } from './inmemory_pubsub.js';
import type { MessageHandler, Message } from './pubsub.js';

/** テスト用: 受信したメッセージを記録するシンプルなハンドラー。 */
function makeHandler(): { handler: MessageHandler; messages: Message[] } {
  const messages: Message[] = [];
  return {
    handler: { handle: async (msg) => { messages.push(msg); } },
    messages,
  };
}

describe('InMemoryPubSub', () => {
  let ps: InMemoryPubSub;

  beforeEach(() => {
    ps = new InMemoryPubSub();
  });

  it('初期状態は uninitialized', async () => {
    expect(await ps.status()).toBe('uninitialized');
  });

  it('init 後は ready になる', async () => {
    await ps.init();
    expect(await ps.status()).toBe('ready');
  });

  it('close 後は closed になりサブスクリプションがクリアされる', async () => {
    await ps.init();
    await ps.close();
    expect(await ps.status()).toBe('closed');
  });

  it('デフォルト name は inmemory-pubsub', () => {
    expect(ps.name).toBe('inmemory-pubsub');
    expect(ps.componentType).toBe('pubsub');
  });

  it('コンストラクタで name を指定できる', () => {
    const named = new InMemoryPubSub('custom-ps');
    expect(named.name).toBe('custom-ps');
  });

  it('metadata は backend=memory を返す', () => {
    expect(ps.metadata()).toEqual({ backend: 'memory' });
  });

  it('publish したメッセージが subscribe で受信できる', async () => {
    await ps.init();
    const { handler, messages } = makeHandler();
    await ps.subscribe('orders', handler);

    await ps.publish('orders', new Uint8Array([1, 2, 3]));

    expect(messages).toHaveLength(1);
    expect(messages[0].topic).toBe('orders');
    expect(messages[0].data).toEqual(new Uint8Array([1, 2, 3]));
  });

  it('メッセージにユニークなIDが付与される', async () => {
    await ps.init();
    const { handler, messages } = makeHandler();
    await ps.subscribe('t', handler);

    await ps.publish('t', new Uint8Array([1]));
    await ps.publish('t', new Uint8Array([2]));

    expect(messages[0].id).toBeTruthy();
    expect(messages[1].id).toBeTruthy();
    expect(messages[0].id).not.toBe(messages[1].id);
  });

  it('メタデータ付きで publish できる', async () => {
    await ps.init();
    const { handler, messages } = makeHandler();
    await ps.subscribe('t', handler);

    await ps.publish('t', new Uint8Array([]), { key: 'value' });

    expect(messages[0].metadata).toEqual({ key: 'value' });
  });

  it('複数のサブスクライバーが同じトピックを受信できる', async () => {
    await ps.init();
    const h1 = makeHandler();
    const h2 = makeHandler();
    await ps.subscribe('events', h1.handler);
    await ps.subscribe('events', h2.handler);

    await ps.publish('events', new Uint8Array([42]));

    expect(h1.messages).toHaveLength(1);
    expect(h2.messages).toHaveLength(1);
  });

  it('別トピックのメッセージは受信しない', async () => {
    await ps.init();
    const { handler, messages } = makeHandler();
    await ps.subscribe('topic-a', handler);

    await ps.publish('topic-b', new Uint8Array([1]));

    expect(messages).toHaveLength(0);
  });

  it('購読者がいないトピックへの publish はエラーにならない', async () => {
    await ps.init();
    await expect(ps.publish('empty', new Uint8Array([]))).resolves.toBeUndefined();
  });

  it('unsubscribe 後はメッセージを受信しない', async () => {
    await ps.init();
    const { handler, messages } = makeHandler();
    const subId = await ps.subscribe('t', handler);
    await ps.unsubscribe(subId);

    await ps.publish('t', new Uint8Array([1]));

    expect(messages).toHaveLength(0);
  });

  it('存在しないサブスクリプションIDの unsubscribe はエラーにならない', async () => {
    await ps.init();
    await expect(ps.unsubscribe('nonexistent-id')).resolves.toBeUndefined();
  });
});

File: regions/system/library/go/building-blocks/inmemory_pubsub_test.go
package buildingblocks

import (
	"context"
	"testing"
	"time"
)

// InMemoryPubSub の Init 前後でステータスが Uninitialized → Ready に遷移することを確認する。
func TestInMemoryPubSub_InitAndStatus(t *testing.T) {
	ps := NewInMemoryPubSub()
	ctx := context.Background()

	if ps.Status(ctx) != StatusUninitialized {
		t.Errorf("expected StatusUninitialized, got %s", ps.Status(ctx))
	}
	if err := ps.Init(ctx, Metadata{}); err != nil {
		t.Fatalf("Init failed: %v", err)
	}
	if ps.Status(ctx) != StatusReady {
		t.Errorf("expected StatusReady, got %s", ps.Status(ctx))
	}
}

// InMemoryPubSub の Name と Version が正しい値を返すことを確認する。
func TestInMemoryPubSub_Name(t *testing.T) {
	ps := NewInMemoryPubSub()
	if ps.Name() != "inmemory-pubsub" {
		t.Errorf("unexpected Name: %q", ps.Name())
	}
	if ps.Version() != "1.0.0" {
		t.Errorf("unexpected Version: %q", ps.Version())
	}
}

// InMemoryPubSub で Publish したメッセージが Subscribe したチャネルから受信できることを確認する。
func TestInMemoryPubSub_PublishSubscribe(t *testing.T) {
	ps := NewInMemoryPubSub()
	ctx := context.Background()
	_ = ps.Init(ctx, Metadata{})

	ch, err := ps.Subscribe(ctx, "events")
	if err != nil {
		t.Fatalf("Subscribe failed: %v", err)
	}

	msg := &Message{Topic: "events", Data: []byte("hello"), ID: "1"}
	if err := ps.Publish(ctx, msg); err != nil {
		t.Fatalf("Publish failed: %v", err)
	}

	select {
	case received := <-ch:
		if received.ID != "1" {
			t.Errorf("expected ID '1', got %q", received.ID)
		}
		if string(received.Data) != "hello" {
			t.Errorf("expected Data 'hello', got %q", received.Data)
		}
	case <-time.After(time.Second):
		t.Fatal("timed out waiting for message")
	}
}

// InMemoryPubSub の Publish が受信メッセージの Timestamp を自動的に設定することを確認する。
func TestInMemoryPubSub_PublishSetsTimestamp(t *testing.T) {
	ps := NewInMemoryPubSub()
	ctx := context.Background()
	_ = ps.Init(ctx, Metadata{})

	ch, _ := ps.Subscribe(ctx, "t")
	msg := &Message{Topic: "t", ID: "x"}
	_ = ps.Publish(ctx, msg)

	select {
	case received := <-ch:
		if received.Timestamp.IsZero() {
			t.Error("expected Timestamp to be set")
		}
	case <-time.After(time.Second):
		t.Fatal("timed out")
	}
}

// InMemoryPubSub が同一トピックの複数サブスクライバー全員にメッセージを配信することを確認する。
func TestInMemoryPubSub_MultipleSubscribers(t *testing.T) {
	ps := NewInMemoryPubSub()
	ctx := context.Background()
	_ = ps.Init(ctx, Metadata{})

	ch1, _ := ps.Subscribe(ctx, "topic")
	ch2, _ := ps.Subscribe(ctx, "topic")

	_ = ps.Publish(ctx, &Message{Topic: "topic", ID: "m"})

	for _, ch := range []<-chan *Message{ch1, ch2} {
		select {
		case <-ch:
		case <-time.After(time.Second):
			t.Fatal("timed out waiting for subscriber")
		}
	}
}

// InMemoryPubSub が異なるトピックのサブスクライバーへメッセージを配信しないことを確認する。
func TestInMemoryPubSub_NoDeliveryToOtherTopics(t *testing.T) {
	ps := NewInMemoryPubSub()
	ctx := context.Background()
	_ = ps.Init(ctx, Metadata{})

	ch, _ := ps.Subscribe(ctx, "other")
	_ = ps.Publish(ctx, &Message{Topic: "events", ID: "m"})

	select {
	case <-ch:
		t.Error("should not receive message for different topic")
	case <-time.After(50 * time.Millisecond):
	}
}

// InMemoryPubSub の Close がステータスを Closed に遷移させサブスクライバーチャネルを閉じることを確認する。
func TestInMemoryPubSub_Close(t *testing.T) {
	ps := NewInMemoryPubSub()
	ctx := context.Background()
	_ = ps.Init(ctx, Metadata{})

	ch, _ := ps.Subscribe(ctx, "t")
	if err := ps.Close(ctx); err != nil {
		t.Fatalf("Close failed: %v", err)
	}
	if ps.Status(ctx) != StatusClosed {
		t.Errorf("expected StatusClosed, got %s", ps.Status(ctx))
	}

	// channel should be closed
	_, ok := <-ch
	if ok {
		t.Error("expected channel to be closed")
	}
}
